src/api/helpers.py

```diff
@@ -0,0 +1,329 @@
+
+    def __init__(self):
+        self.id_map = {}
+
+    def create_key(self, id_obj: dict) -> str:
+        """Create a unique string key from an id_map object."""
+        return f"{id_obj['source_system']}|{id_obj['source_value']}|{id_obj['source_desc']}|{id_obj['target_desc']}"
+
+    def store_id(self, id_obj: dict, actual_id: int):
+        """Store the mapping between an id_map object and an actual database ID."""
+        key = self.create_key(id_obj)
+        self.id_map[key] = actual_id
+
+    def get_id(self, id_obj: dict) -> int | None:
+        """Retrieve the actual database ID for an id_map object."""
+        key = self.create_key(id_obj)
+        return self.id_map.get(key)
+
+
+# ==============================================================================
+# Create object in OMOP
+# ==============================================================================
+async def create_record(
+    session: AsyncSession,
+    id_mapper: IdMapper,
+    record_field: dict,
+    table_name: str,
+) -> dict:
+    """
+    Function to process and create a single OMOP record.
+    """
+    config = TABLE_CONFIG[table_name]
+    record_data = record_field.get("omop_record", {})
+
+    # 1. Extract the temporary ID object for the primary key, if it exists
+    pk_field = config.get("pk")
+    id_obj = None
+    if pk_field:
+        id_obj = record_data.pop(pk_field, None)
+
+    # 2. Resolve all foreign keys using the id_mapper
+    for fk_field_name in config["fk_map"]:
+        if fk_field_name in record_data:
+            fk_id_obj = record_data[fk_field_name]
+            actual_fk_id = id_mapper.get_id(fk_id_obj)
+            if actual_fk_id is None:
+                raise ProblemException(
+                    status=400,
+                    title="Bad Request",
+                    detail=(
+                        f"Could not resolve foreign key for '{fk_field_name}' in table '{table_name}'. "
+                        f"Ensure the referenced entity is created first. "
+                        f"Missing reference: {fk_id_obj}"
+                    ),
+                )
+            record_data[fk_field_name] = actual_fk_id
+
+    # 3. Call the specific create function for this table
+    create_function = config["create_func"]
+    new_record = await create_function(session, record_data)
+
+    # 4. Store the mapping for the new primary key
+    if pk_field and id_obj:
+        new_pk_value = new_record[pk_field]
+        id_mapper.store_id(id_obj, new_pk_value)
+
+    # 5. Prepare the record for the final return object
+    new_record["omop_table"] = table_name
+    return new_record
+
+
+# ==============================================================================
+# Insert donor with clinical data
+# ==============================================================================
+async def ingest_donor_with_clinical_data(
+    session: AsyncSession, donor_data: dict, return_obj: list
```

Comment:
Why is return_obj given to the function here? It seems like it's an empty list every time, according to `handle_single_donor_data_ingestion`, and it's strange that this function is not idempotent. 

Recommend:
1. switching it so that `return_obj` is instantiated as a list inside this function, and returned. Callers can merge the list with any existing ingests if they want
2. Renaming `return_obj` to `return_objs`, as it's a bit confusing to see `*_obj` refer to a list instead of a dict

Reply:
Yeah, this is valid. I'll fix it. The function used to process all the data before Justin split it each donor with clinical data. I refactored and overlooked some leftover logic.